presto-main/src/main/java/com/facebook/presto/operator/RowNumberOperator.java

```diff
@@ -195,12 +200,7 @@ public boolean needsInput()
             // Check if single partition is done
             return partitionRowCount.get(0) < maxRowsPerPartition.get() && !finishing && inputPage == null;
         }
-        return !finishing && inputPage == null;
```

Comment:
any reason why you didn't add `unfinishedWork == null` as another condition to `isFinished()`?